### PR TITLE
Handle reporters with absolute path as output path 

### LINKED ISSUE #48: Cannot specify report with absolute path
If you try to specify `json` or `pmd` with an absolute path, it says "Error while writing the <XML|JSON> object to file."

Example:
```
taylor -r json:`pwd`/report.json 
```
*Note*: It works if you specify a relative path.

### PR DESCRIPTION
**Pull request by mihai8804858**
This PR will fix #48 issue.

Also, slightly refactored reporter coordinators.

### CODE REVIEW

File: TaylorIntegrationTests/ReporterComparator.swift [hunk 1]
```diff
@@ -33,14 +33,14 @@ func !=(lhs: [String:AnyObject], rhs: [String:AnyObject]) -> Bool {
 
 class ReporterComparator {
     
-    func compareReporters(_ firstReporterPath: String, secondReporterPath: String) -> Bool {
-        var violations1 = JSONToViolationParser().parseFile(firstReporterPath)
+    func compareReporter(atPath path1: String, withReporterAtPath path2: String) -> Bool {
```
Comment (lexorus) [inline TaylorIntegrationTests/ReporterComparator.swift]: This is the function that gets two paths, then parses this files to get violations, then compares each violation through iteration. 
All this is happening in `compareReporter(atPath:withReporterAtPath:)` function of `ReporterComparator` class.
The Reporter word leads to a delusion here(we also have reporters in this project).
Can you think about a better name, like `compareViolations`, or for a better place/implementation of this functionality(extension)?

File: TaylorFramework/Modules/temper/Output/OutputCoordinator.swift [hunk 2]
```diff
@@ -19,11 +19,15 @@ final class OutputCoordinator {
...
         self.reporters = reporters
         for reporter in reporters {
-            if reporter.fileName.isEmpty && (reporter.concreteReporter as? XcodeReporter) == nil { continue }
-            let path = (filePath as NSString).appendingPathComponent(reporter.fileName)
-            reporter.coordinator().writeViolations(violations, atPath: path)
+            writeOutput(with: violations, reporter: reporter)
         }
     }
+    
+    private func writeOutput(with violations: [Violation], reporter: Reporter) {
```
Comment (lexorus) [inline TaylorFramework/Modules/temper/Output/OutputCoordinator.swift]: I think we can get rid of the `Output` word here.

**Review by lexorus — COMMENTED** [review]
The mentioned issue itself is fixed.

**Comment by lexorus** [comment]
Also, there is a Typo in commit:
Handle absolute paths when _**specifie**_ reporter's output path in arguments

File: TaylorIntegrationTests/ReporterComparator.swift (hunk 1 above)
Comment (mihai8804858) [reply TaylorIntegrationTests/ReporterComparator.swift]: This class takes 2 paths to reports and compare them. I'm not interested how it compares the reports, it's his problem. Also, If I would see `compareViolations` function I would think that I should give it an array of `Violation`.

Comment (lexorus) [reply TaylorIntegrationTests/ReporterComparator.swift]: Then, maybe will be better to name it `compareReport(atPath:withReportAtPath:)` (**report** and not **reporter**)?

Comment (mihai8804858) [reply TaylorIntegrationTests/ReporterComparator.swift]: Already done that.

**Comment by thelvis4** [comment]
@mihai8804858 Tests are failing 😢

**Comment by mihai8804858** [comment]
@thelvis4 Don't know why 😢 Have you tried to run the tests on your machine? On my machine and on @lexorus's the tests are passing.

**Comment by thelvis4** [comment]
@mihai8804858 @lexorus Checks are failing because code coverage decreased. 😢

File: TaylorFramework/Modules/temper/Output/PMDCoordinator.swift [hunk 3]
```diff
@@ -15,36 +15,41 @@ final class PMDCoordinator: WritingCoordinator {
...
-            }
-            fileElement.addAttribute(attribute)
-            let _ = violations.filter({ $0.path == filePath }).map({ fileElement.addChild($0.toXMLElement()) })
-            if let root = xml.rootElement() {
-                root.addChild(fileElement)
-            }
+    private func addElements(to xml: XMLDocument, from violations: [Violation]) {
+        for path in filePaths(from: violations) {
+            guard let fileElement = xmlElement(withFilePath: path) else { continue }
+            violations.filter { $0.path == path }
```
Comment (thelvis4) [inline TaylorFramework/Modules/temper/Output/PMDCoordinator.swift]: It seems like we do a lot of iteration here. A big number of violations can lead to a computation time growth.
One way we can reduce the numbers of loops is by iterating through violations (just once), group them into a dictionary that uses paths as keys:
```swift
private typealias ViolationsMap = [Path: [Violation]]

final class PMDCoordinator: WritingCoordinator {
    
    func writeViolations(_ violations: [Violation], atPath path: String) {
        FileManager().removeFileAtPath(path)
        let map = mapViolations(violations)
        let xml = generateXML(from: map)
        let xmlData = xml.xmlData(withOptions: Int(XMLNode.Options.nodePrettyPrint.rawValue))
        do {
            try xmlData.write(to: URL(fileURLWithPath: path), options: .atomic)
        } catch let error {
            let message = "Error while writing XML object at path: " + path + "\n" +
                "Reason: " + error.localizedDescription
            Printer(verbosityLevel: .error).printError(message)
        }
    }
    
    private func mapViolations(_ violations: [Violation]) -> ViolationsMap {
        return violations.reduce([Path : [Violation]]()) { result, violation in
            let path = violation.path
            
            // Add violation to the group with the same path
            var violations = result[path] ?? []
            violations.append(violation)
            
            // Update the result
            var nextResult = result
            nextResult[path] = violations
            
            return result
        }
    }
    
    private func generateXML(from violationsMap: ViolationsMap) -> XMLDocument {
        let xml = XMLDocument(rootElement: XMLElement(name: "pmd"))
        xml.version = "1.0"
        xml.characterEncoding = "UTF-8"
        
        let fileElements = violationsMap.map(generateFileElement)
        xml.rootElement()?.addChildren(fileElements)
        
        return xml
    }
    
    private func generateFileElement(path: Path, violations: [Violation]) -> XMLElement {
        let element = XMLElement(name: "file")
        let attribute = XMLNode(kind: .attribute)
        attribute.setValue(path, forKey: "name")
        element.addAttribute(attribute)
        
        violations.forEach { element.addChild($0.toXMLElement()) }
        
        return element
    }
}

extension XMLElement {
    func addChildren(_ children: [XMLElement]) {
        children.forEach { child in
            addChild(child)
        }
    }
}
```
What do you think?

File: TaylorIntegrationTests/TaylorIntegrationTests.swift [hunk 4]
```diff
@@ -17,7 +17,7 @@ enum IntegrationTestsError: Error {
 
 class TaylorIntegrationTests: QuickSpec {
     
-    var runTaskPath : String {
+    var taylorExecutablePath : String {
```
Comment (thelvis4) [inline TaylorIntegrationTests/TaylorIntegrationTests.swift]: How about `executableFilePath`?
If we rename the tool, we don't want to be needed to make too many changes.

File: TaylorIntegrationTests/TaylorIntegrationTests.swift [hunk 5]
```diff
@@ -88,6 +88,23 @@ class TaylorIntegrationTests: QuickSpec {
         }
     }
     
+    fileprivate func runTaylorWithArguments(_ arguments: String...) {
```
Comment (thelvis4) [inline TaylorIntegrationTests/TaylorIntegrationTests.swift]: `runWithArguments`?

File: TaylorIntegrationTests/TaylorIntegrationTests.swift [hunk 6]
```diff
@@ -116,20 +133,27 @@ class TaylorIntegrationTests: QuickSpec {
...
-                runTask.waitUntilExit()
-                let reporterPath = self.analyzeFilesPath.stringByAppendingPathComponent("taylor_report.json")
-                expect(ReporterComparator().compareReporters(self.resultFilePath, secondReporterPath: reporterPath)).to(beTrue())
-                do {
-                    if FileManager.default.fileExists(atPath: reporterPath) {
-                        try FileManager.default.removeItem(atPath: reporterPath)
-                    }
-                } catch let error {
-                    print(error)
+            context("when specifie report with relative path") {
```
Comment (thelvis4) [inline TaylorIntegrationTests/TaylorIntegrationTests.swift]: Typo here at `specifie`

File: TaylorIntegrationTests/TaylorIntegrationTests.swift [hunk 7]
```diff
@@ -116,20 +133,27 @@ class TaylorIntegrationTests: QuickSpec {
...
+                                                "-r", "json:taylor_report.json")
+                    let reporterPath = self.analyzeFilesPath.stringByAppendingPathComponent("taylor_report.json")
+                    let comparisonResult = JSONReportComparator().compareReport(atPath: self.resultFilePath,
+                                                                                withReportAtPath: reporterPath)
+                    expect(comparisonResult).to(beTrue())
+                    self.removeFile(atPath: reporterPath)
+                }
+            }
+            
+            context("when specifie report with absolute path") {
```
Comment (thelvis4) [inline TaylorIntegrationTests/TaylorIntegrationTests.swift]: Same tipo (`specifie`) here 😉

**Review by thelvis4 — CHANGES_REQUESTED** [review]

File: TaylorIntegrationTests/TaylorIntegrationTests.swift (hunk 7 above)
Comment (mihai8804858) [reply TaylorIntegrationTests/TaylorIntegrationTests.swift]: Typo here: `tipo` -> `typo` 😉

File: TaylorFramework/Modules/temper/Output/PMDCoordinator.swift (hunk 3 above)
Comment (mihai8804858) [reply TaylorFramework/Modules/temper/Output/PMDCoordinator.swift]: I think you did a good job! 👍

File: TaylorIntegrationTests/TaylorIntegrationTests.swift (hunk 7 above)
Comment (thelvis4) [reply TaylorIntegrationTests/TaylorIntegrationTests.swift]: 🤣

**Comment by thelvis4** [comment]
@mihai8804858 There is a typo in the commit message. Could you please fix it?

File: TaylorFrameworkTests/TemperTests/NSFileManagerTests.swift [hunk 8]
```diff
@@ -26,6 +36,18 @@ class NSFileManagerTests: QuickSpec {
                 manager.removeFileAtPath(filePath)
                 expect(manager.fileExists(atPath: filePath)).to(beFalse())
             }
+            
+            it("should't remove file at path if it's not deleatable") {
```
Comment (thelvis4) [inline TaylorFrameworkTests/TemperTests/NSFileManagerTests.swift]: There is no such word `deleatable`. You can use `deletable` or `delible`, but I recommend to use  `removable` here as it's less pretentious.

Comment (thelvis4) [reply TaylorFrameworkTests/TemperTests/NSFileManagerTests.swift]: Typo: `should't`

**Review by thelvis4 — CHANGES_REQUESTED** [review]

File: TaylorFrameworkTests/TemperTests/NSFileManagerTests.swift (hunk 8 above)
Comment (mihai8804858) [reply TaylorFrameworkTests/TemperTests/NSFileManagerTests.swift]: There is a method in 'FileManager' called 'isDeleatableFileAtPath', I throught there is such word. 🙂

**Comment by mihai8804858** [comment]
@thelvis4 Fixed typos.

**Comment by thelvis4** [comment]
@mihai8804858  It seems like we introduced a bug here.
Before we had
```xml
<file name="/Some/path/to/file.swift">
```
but now we have
```xml
<file /Some/path/to/file.swift="">
```
Could you please investigate and also write some tests to make sure the bug is fixed and we won't have regressions?

File: TaylorFrameworkTests/TemperTests/NSFileManagerTests.swift [hunk 9]
```diff
@@ -26,6 +36,18 @@ class NSFileManagerTests: QuickSpec {
...
+                let path = manager.currentDirectoryPath as NSString
+                let filePath = path.appendingPathComponent("file.txt") as String
+                manager.createFile(atPath: filePath, contents: nil, attributes: nil)
+                expect(manager.fileExists(atPath: filePath)).to(beTrue())
+                manager.removeFileAtPath(filePath)
+                expect(manager.fileExists(atPath: filePath)).to(beTrue())
+                FileManager.default.removeFileAtPath(filePath)
+            }
+            
             it("should't remove directory at path") {
```
Comment (thelvis4) [inline TaylorFrameworkTests/TemperTests/NSFileManagerTests.swift]: It seems like we have the same typo here: `should't`

**Review by thelvis4 — CHANGES_REQUESTED** [review]

**Comment by mihai8804858** [comment]
@thelvis4 Done.

**Review by thelvis4 — APPROVED** [review]

**Comment by thelvis4** [comment]
@mihai8804858 How about the issue with `file` element? It's still reproducing to me.

**Comment by mihai8804858** [comment]
Oops, I missed that comment. 😕

**Comment by thelvis4** [comment]
@mihai8804858 Thanks for working on this! 👍 
You just made `Taylor` better! 👧 😊

**Pull request merged**